src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Endpoints/RazorCustomMessageTarget_UpdateCSharpBuffer.cs

```diff
@@ -24,20 +25,34 @@ public async Task UpdateCSharpBufferAsync(UpdateBufferRequest request, Cancellat
             throw new ArgumentNullException(nameof(request));
         }
 
-        await _joinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-
-        await UpdateCSharpBufferCoreAsync(request, cancellationToken);
+        // We're going to try updating the C# buffer, and it has to happen on the UI thread. That is a single shared resource
+        // that can only be updated on a specific thread, and so we can easily hit contention. In particular with provisional
+        // completion we can end up with a state where we are waiting for completion to finish before we can update the document
+        // but other features are waiting on us to update the document with some changes. Normally this is fine, and we handle it
+        // with our sync system and simply cancel the task and let the next one in. When updating buffers from the server though
+        // we can't do that, as the server assumes we can always apply the update. Our only option here is to keep trying until
+        // we get a successful update.
+        // It's worth noting we only try again specifically for provisional completion, so we don't get random deadocks.
+        var tryAgain = true;
+        while (tryAgain)
+        {
+            _logger.LogDebug($"Trying a call to UpdateCSharpBufferCoreAsync for v{request.HostDocumentVersion}");
+            tryAgain = await UpdateCSharpBufferCoreAsync(request, cancellationToken).ConfigureAwait(false);
+        }
     }
 
     // Internal for testing
```

Comment:
FWIW, it appears that this "internal for testing" method might not be justified since it actually does most of the work of `UpdateCSharpBufferAsync`. Should tests be updated to call `UpdateCSharpBufferAsync` so this can be private?